src/transformStateWithClones.js

```diff
@@ -7,7 +7,25 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    if (action.type === 'clear') {
+      currentState = {};
+    } else if (action.type === 'addProperties') {
+      currentState = { ...currentState, ...action.extraData };
+    } else if (action.type === 'removeProperties') {
+      for (const key of action.keysToRemove) {
+        delete currentState[key];
+      }
+    }
+
+    stateHistory.push({ ...currentState });
+  }
+
+  return stateHistory;
```

Comment:
The variable name 'currentState' is good, but it's better to push a copy of the object to the 'stateHistory' array at the end of the loop cycle, outside of the switch block, after the current action is processed to avoid repeating yourself.